imx_pwm_led.c

```diff
@@ -0,0 +1,2069 @@
+	char *dev_name;
+	/** PID of owner */
+	pid_t owner_pid;
+	/** Currently open object */
+	enum open_object open_object;
+	/** Index of currently open object */
+	uint open_object_idx;
+	/** True if writing is ongoing */
+	int writing;
+	/** True if adaptive mode is enabled */
+	int adaptive;
+	/** Index of the currently playing fade, or -1 if not playing one */
+	int playing_fade_idx;
+	/** PWM FIFO sample period in microseconds */
+	uint pwm_fifo_period_us;
+};
+
+/** Fade struct */
+struct led_fade {
+	/** Buffer of u16 samples */
+	void *buf;
+	/** Physical address of buffer for the SDMA */
+	dma_addr_t buf_phys;
+	/** Length of fade data in **bytes** */
+	uint len;
+	/** Lock to guard access */
+	struct mutex lock;
+	/** PID of owner */
+	pid_t owner_pid;
+	/** Direction */
+	enum fade_dir dir;
+	/** Atomic reference counter for LEDs playing the fade */
+	atomic_t player_count;
+};
+
+/** Cue action struct */
+struct led_cue_action {
+	/** Index of LED */
+	u8 led_idx;
+	/** Type, see PWM_LED_CUE_ACTION_TYPE_* */
+	u8 type;
+	/** Value */
+	u16 val;
+};
+
+/** Cue struct */
+struct led_cue {
+	/** Buffer of actions */
+	struct led_cue_action *actions;
+	/** Length of action data in **bytes** */
+	uint len;
+	/** Lock to guard access */
+	struct mutex lock;
+	/** PID of owner */
+	pid_t owner_pid;
+};
+
+/** Module global data struct */
+struct glb_data {
+	/** Array of LEDs */
+	struct imx_pwm_led **leds;
+	/** Array of fades */
+	struct led_fade *fades;
+	/** Array of cues */
+	struct led_cue *cues;
+	/** Buffer of duty cycles [(max_cues + 1) * max_leds] */
+	u16 *duty_buf;
+	/** Physical address of duty buffer for the SDMA */
+	dma_addr_t duty_buf_phys;
+	/** Number of configured LEDs */
+	uint num_leds;
+	/** Pointer to the timer used to drive the DMA engine */
+	struct epit *epit;
+	/** Pointer to the SDMA engine */
+	struct sdma_engine *sdma;
+	/** SDMA script load address (SDMA address in words) */
+	u32 sdma_script_origin;
+};
+
+/*******************************************************************************
+ * VARIABLES & CONSTANTS
+ ******************************************************************************/
+
+/* clang-format off */
+/** Module Parameters */
+static u32 sample_rate      = 100; /* Fade sample rate in Hz */
+static u32 pwm_period       = 24000; /* 24 MHz IPG clk / 24000 = 1 kHz */
+static u32 pwm_prescaler    = 0; /* N-1 value, allowed range: 0 to 4095 */
+static u32 pwm_invert       = 0; /* 0=non-inverted, 1=inverted */
+static u32 pwm_phase_offset = 0; /* PWM channel phase offset in microseconds */
+static u32 pwm_repeat       = 0; /* 2^N value, allowed range: 0 to 3  */
+static u32 max_fades        = 16;
+static u32 max_cues         = 16;
+static u32 max_leds         = 8; /* imx6ul has 8 PWMs */
+static u32 max_fade_size    = 4096; /* DMA memory block size is 4KB */
+static u32 sdma_priority    = 5; /* Allowed range: 1 to 6 */
+static u32 perm_mode        = PERM_MODE_PID_OWNER; /* Permission mode */
```

Comment:
Yes exactly the `module_param` macro makes the variable visible as a module parameter with exactly that name.
Yes it would be nice to have typedef and a `_t` suffix, but the linux coding guideline says not to use typedef, rather enum/struct before the tag name.